release/nightly_tests/dataset/map_batches_benchmark.py

```diff
@@ -0,0 +1,128 @@
+        for batch_size in batch_sizes:
+            # TODO(chengsu): Investigate why NumPy with batch_size less than 512,
+            # took forever to finish.
+            if batch_format == "numpy" and batch_size is not None and batch_size < 512:
+                continue
+
+            num_calls = 2
+            test_name = f"map-batches-{batch_format}-{batch_size}-{num_calls}-default"
+            benchmark.run(
+                test_name,
+                map_batches,
+                input_ds=input_ds,
+                batch_format=batch_format,
+                batch_size=batch_size,
+                num_calls=num_calls,
+            )
+            test_name = f"map-batches-{batch_format}-{batch_size}-{num_calls}-lazy"
+            benchmark.run(
+                test_name,
+                map_batches,
+                input_ds=lazy_input_ds,
+                batch_format=batch_format,
+                batch_size=batch_size,
+                num_calls=num_calls,
+            )
+
+    # Test multiple calls of map_batches.
+    for batch_format in batch_formats:
+        for num_calls in num_calls_list:
+            for compute in ["tasks", ActorPoolStrategy(1, 1)]:
+                batch_size = 4096
+                if compute == "tasks":
+                    compute_strategy = "task"
+                else:
+                    compute_strategy = "actor"
+
+                test_name = (
+                    f"map-batches-{batch_format}-{batch_size}-{num_calls}-"
+                    f"{compute_strategy}-default"
+                )
+                benchmark.run(
+                    test_name,
+                    map_batches,
+                    input_ds=input_ds,
+                    batch_format=batch_format,
+                    batch_size=batch_size,
+                    compute=compute,
+                    num_calls=num_calls,
+                )
+                test_name = (
+                    f"map-batches-{batch_format}-{batch_size}-{num_calls}-"
+                    f"{compute_strategy}-lazy"
+                )
+                benchmark.run(
+                    test_name,
+                    map_batches,
+                    input_ds=lazy_input_ds,
+                    batch_format=batch_format,
+                    batch_size=batch_size,
+                    compute=compute,
+                    num_calls=num_calls,
+                )
+
+    # Test different batch formats of map_batches.
+    for current_format in ["pyarrow", "pandas", "numpy"]:
+        new_input_ds = input_ds.map_batches(
+            lambda ds: ds, batch_format=current_format, batch_size=None
+        ).fully_executed()
+        for new_format in ["pyarrow", "pandas", "numpy"]:
+            test_name = f"map-batches-{current_format}-to-{new_format}"
+            benchmark.run(
+                test_name,
+                map_batches,
+                input_ds=new_input_ds,
+                batch_format=new_format,
+                batch_size=None,
```

Comment:
Batching can cause this conversion to be more or less expensive, e.g. batch sizes that are not aligned with upstream chunking can result in a lot of concatenation and slicing. Might be worth capturing this.

Reply:
updated.